src/jit/compiler.h

```diff
@@ -3108,6 +3108,7 @@ class Compiler
     static int ivalOfHWIntrinsic(NamedIntrinsic intrinsic);
     unsigned simdSizeOfHWIntrinsic(NamedIntrinsic intrinsic, CORINFO_SIG_INFO* sig);
     static int numArgsOfHWIntrinsic(NamedIntrinsic intrinsic);
+    static int numArgsOfHWIntrinsic(NamedIntrinsic intrinsic, GenTreeHWIntrinsic* node);
```

Comment:
Missed the conversation here. When is one overload called vs the other?

Reply:
When `numArgs` is -1 in table and we need exact `numArgs` to handle node. There are 3 cases one in each: `lsraxarch.cpp`, lowerxarch.cpp` and `hwintrinsiccodegenxarch.cpp` when new overload is used.

Reply:
Given that the new overload always calls into the original first, why not have a single overload to reduce confusion?

The original overload can be directly inlined (given how small it is).

Reply:
Actually the code of original overload is used verbatim during first check see:

https://github.com/dotnetrt/coreclr/blob/d4bb1d346291581a012b336aac12f46b66e2ae3f/src/jit/hwintrinsicxarch.cpp#L253

Passing 2 args to call and conditional return make new overload around one - two order of magnitudes slower than original one since original overload should be inlined by C++ compilers anyway.

Reply:
I suggest to remove the 1-arg overload.

Reply:
Yes, I think I suggested a single overload in my earlier comments (prior to @fiigii's excellent suggestion to avoid using the base type altogether). You could have `node` default to `nullptr` and then add a `noway_assert` that it is non-null if `numArgs == -1`, followed conservatively by a check that it is non-null before doing the rest of the method. If by mistake a caller fails to pass a non-null `node` argument when they should, the behavior will be the same "as if" the 1-arg overload had been called. And with the null argument, the C++ compiler should be expected to do the inlining and the elimination of the non-null path.
In any event, we want to be very careful about pre-optimizing code at the expense of clarity and conciseness.

Reply:
> Passing 2 args to call and conditional return make new overload around one - two order of magnitudes slower than original one since original overload should be inlined by C++ compilers anyway.

@4creators, but under what scenario is calling the original one valid? It seems like, given that certain table entries can have -1 for the numArgs, it is almost always appropriate to only call the two overload version, so it is the only one we should have.

Reply:
The arguments are strong and I agree that it would be best approach. Will fix.